### PR TITLE
Resources for GitHub actions

### PR DESCRIPTION
**Pull request by MancunianSam**
- Roles and resources for github actions
- Update IAM policy to use WAFV2
- Add policy to deploy lambda
- Use new secrets
- Ran terraform fmt
- Add role to deploy lambda
- Use the correct role name
- Add permissions to lambda deploy role
- Set the env vars correctly
- Add params for consignment api repo
- Allow access for all tdr repos

### CODE REVIEW

File: root.tf [hunk 1]
```diff
@@ -673,6 +673,32 @@ module "bastion_role" {
   policy_attachments = {}
 }
 
+module "update_waf_and_security_groups_policy" {
+  source        = "./tdr-terraform-modules/iam_policy"
+  name          = "UpdateWAFAndSecurityGroupsPolicy${title(local.environment)}"
```
Comment (TomJKing) [inline root.tf:678]: Needs `TDR` add to the beginning of the name maybe?

File: root.tf [hunk 2]
```diff
@@ -673,6 +673,32 @@ module "bastion_role" {
...
+  policy_string = templatefile("${path.module}/templates/iam_policy/update_waf_and_security_groups_policy.json.tpl", { rule_group_arn = module.waf.rule_group_arn, ip_set_arn = module.waf.ip_set_arn, account_id = data.aws_caller_identity.current.account_id, security_group_id = module.frontend.alb_security_group_id })
+}
+
+module "update_waf_and_security_groups_role" {
+  source             = "./tdr-terraform-modules/iam_role"
+  assume_role_policy = templatefile("${path.module}/templates/iam_role/github_assume_role.json.tpl", { account_id = data.aws_caller_identity.current.account_id, repo_name = "tdr-e2e-tests" })
+  common_tags        = local.common_tags
+  name               = "UpdateWAFAndSecurityGroupsRole${title(local.environment)}"
```
Comment (TomJKing) [inline root.tf:686]: Needs `TDR` add to the beginning of the name maybe?

File: root.tf [hunk 3]
```diff
@@ -733,3 +799,34 @@ module "create_bulk_users_bucket" {
...
+    TITLE_STAGE            = title(local.environment)
+    ACCOUNT_NUMBER         = data.aws_caller_identity.current.account_id
+    MANAGEMENT_ACCOUNT     = data.aws_ssm_parameter.mgmt_account_number.value
+    SLACK_FAILURE_WORKFLOW = data.aws_ssm_parameter.slack_failure_workflow.value
+    SLACK_SUCCESS_WORKFLOW = data.aws_ssm_parameter.slack_success_workflow.value
+    WORKFLOW_PAT           = data.aws_ssm_parameter.workflow_pat.value
+  }
+}
+
+module "deploy_lambda_github_actions_policy" {
```
Comment (TomJKing) [inline root.tf:818]: Maybe rename slightly to provide a consistent naming convention for the github stuff: `github_actions_deploy_lambda_policy` ?

File: root.tf [hunk 4]
```diff
@@ -673,6 +673,32 @@ module "bastion_role" {
...
+  source             = "./tdr-terraform-modules/iam_role"
+  assume_role_policy = templatefile("${path.module}/templates/iam_role/github_assume_role.json.tpl", { account_id = data.aws_caller_identity.current.account_id, repo_name = "tdr-e2e-tests" })
+  common_tags        = local.common_tags
+  name               = "UpdateWAFAndSecurityGroupsRole${title(local.environment)}"
+  policy_attachments = {
+    update_policy = module.update_waf_and_security_groups_policy.policy_arn
+  }
+}
+
+module "update_ecs_role" {
```
Comment (TomJKing) [inline root.tf:692]: Could the module be renamed slightly to make it clearer what this is: `github_update_ecs_role`

File: root.tf [hunk 5]
```diff
@@ -691,6 +717,46 @@ module "s3_vpc_endpoint" {
...
 
+module "github_oidc_provider" {
+  source      = "./tdr-terraform-modules/identity_provider"
+  audience    = "sts.amazonaws.com"
+  thumbprint  = "6938fd4d98bab03faadb97b34396831e3780aea1"
+  url         = "https://token.actions.githubusercontent.com"
+  common_tags = local.common_tags
+}
+
+module "consignment_api_github_environment" {
```
Comment (TomJKing) [inline root.tf:728]: Maybe rename slightly to make clearer what this module is: `github_consignment_api_environment`

File: root.tf [hunk 6]
```diff
@@ -691,6 +717,46 @@ module "s3_vpc_endpoint" {
...
+    TITLE_STAGE            = title(local.environment)
+    ACCOUNT_NUMBER         = data.aws_caller_identity.current.account_id
+    MANAGEMENT_ACCOUNT     = data.aws_ssm_parameter.mgmt_account_number.value
+    SLACK_FAILURE_WORKFLOW = data.aws_ssm_parameter.slack_failure_workflow.value
+    SLACK_SUCCESS_WORKFLOW = data.aws_ssm_parameter.slack_success_workflow.value
+    WORKFLOW_PAT           = data.aws_ssm_parameter.workflow_pat.value
+  }
+}
+
+module "e2e_tests_github_environment" {
```
Comment (TomJKing) [inline root.tf:743]: Maybe rename slightly to make clearer what this module is: `github_e2e_tests_environment`

File: root.tf [hunk 7]
```diff
@@ -733,3 +799,34 @@ module "create_bulk_users_bucket" {
   lambda_notification = true
   lambda_arn          = module.create_keycloak_users_s3_lambda.create_keycloak_users_s3_lambda_arn
 }
+
+module "keycloak_user_management_github_environment" {
```
Comment (TomJKing) [inline root.tf:803]: Maybe rename this module slightly to make this clearer what it is: `github_keycloak_user_management_environment`

File: root.tf [hunk 8]
```diff
@@ -673,6 +673,32 @@ module "bastion_role" {
   policy_attachments = {}
 }
 
+module "update_waf_and_security_groups_policy" {
```
Comment (TomJKing) [inline root.tf:676]: I'd be tempted to move the specific github modules into a separate root file by themselves: `root_github` ?

File: root.tf [hunk 9]
```diff
@@ -714,7 +780,7 @@ module "create_keycloak_users_s3_lambda" {
   vpc_id                         = module.shared_vpc.vpc_id
   lambda_create_keycloak_user_s3 = true
   private_subnet_ids             = module.backend_checks_efs.private_subnets
-  s3_bucket_arn                  = module.create_bulk_users_bucket.s3_bucket_arn
+  s3_bucket_arn                  = "arn:aws:s3:::tdr-create-bulk-keycloak-users-intg"
```
Comment (TomJKing) [inline root.tf:783]: Should the ARN have a hardcoded TDR environment `intg` at the end?

File: templates/iam_policy/deploy_lambda_github_actions.json.tpl [hunk 10]
```diff
@@ -0,0 +1,28 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Action": [
```
Comment (TomJKing) [inline templates/iam_policy/deploy_lambda_github_actions.json.tpl:7]: Maybe order alphabetically?

File: templates/iam_policy/github_sbt_dependencies_policy.json.tpl [hunk 11]
```diff
@@ -0,0 +1,20 @@
+{
+  "Statement": [
+    {
+      "Action": [
```
Comment (TomJKing) [inline templates/iam_policy/github_sbt_dependencies_policy.json.tpl:4]: Maybe order alphabetically?

File: templates/iam_policy/github_sbt_dependencies_policy.json.tpl [hunk 12]
```diff
@@ -0,0 +1,20 @@
...
+      "Resource": [
+        "arn:aws:s3:::tdr-snapshots-mgmt/*",
+        "arn:aws:s3:::tdr-snapshots-mgmt",
+        "arn:aws:s3:::tdr-releases-mgmt/*",
+        "arn:aws:s3:::tdr-releases-mgmt"
+      ],
+      "Sid": ""
+    }
+  ],
+  "Version": "2012-10-17"
```
Comment (TomJKing) [inline templates/iam_policy/github_sbt_dependencies_policy.json.tpl:19]: Maybe move the `Version` field to beginning of policy as per the other policy declarations?

File: templates/iam_policy/update_waf_and_security_groups_policy.json.tpl [hunk 13]
```diff
@@ -0,0 +1,31 @@
+{
+  "Version": "2012-10-17",
+  "Statement": [
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Action": [
```
Comment (TomJKing) [inline templates/iam_policy/update_waf_and_security_groups_policy.json.tpl:7]: Maybe order alphabetically?

File: templates/iam_policy/update_waf_and_security_groups_policy.json.tpl [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+        "${ip_set_arn}",
+        "${rule_group_arn}",
+        "arn:aws:wafv2:eu-west-2:${account_id}:REGIONAL/regexpatternset/*/*",
+        "arn:aws:ec2:eu-west-2:${account_id}:security-group/${security_group_id}"
+      ]
+    },
+    {
+      "Sid": "",
+      "Effect": "Allow",
+      "Action": [
```
Comment (TomJKing) [inline templates/iam_policy/update_waf_and_security_groups_policy.json.tpl:23]: Maybe order alphabetically?

File: versions.tf [hunk 15]
```diff
@@ -4,6 +4,10 @@ terraform {
       source  = "hashicorp/aws"
       version = "3.72.0"
     }
+    github = {
+      source  = "integrations/github"
+      version = "~> 4.0"
```
Comment (TomJKing) [inline versions.tf:9]: That's a bold move...

**Review by TomJKing — COMMENTED** [review]
Looks good. Just minor comments around renaming etc.

File: root.tf (hunk 8 above)
Comment (MancunianSam) [reply root.tf:676]: Yeah that makes sense, the main root file is getting a bit big.

File: root.tf (hunk 9 above)
Comment (MancunianSam) [reply root.tf:783]: No it shouldn't, no idea why I did this. I've put the module back.

File: versions.tf (hunk 15 above)
Comment (MancunianSam) [reply versions.tf:9]: Terraform solves all problems :smile:

File: root_github.tf [hunk 16]
```diff
@@ -0,0 +1,102 @@
+module "github_oidc_provider" {
+  source      = "./tdr-terraform-modules/identity_provider"
+  audience    = "sts.amazonaws.com"
+  thumbprint  = "6938fd4d98bab03faadb97b34396831e3780aea1"
```
Comment (TomJKing) [inline root_github.tf:4]: Assuming this is OK to be exposed in the repo? Would this be better in the tdr-configurations repo?

**Review by TomJKing — COMMENTED** [review]
Looks good. Just one minor comment. Probably being too fussy, but thought best to ask.

File: root_github.tf (hunk 16 above)
Comment (MancunianSam) [reply root_github.tf:4]: It's a public thumbprint GitHub provide. I got it from here https://github.blog/changelog/2022-01-13-github-actions-update-on-oidc-based-deployments-to-aws/ so I think it's ok.

Comment (TomJKing) [reply root_github.tf:4]: Fair enough.

**Review by TomJKing — APPROVED** [review]
Looks good.

**Pull request merged**